README.md
[![NPM version](https://badge.fury.io/js/grunt-node-version.png)](https://npmjs.org/package/grunt-node-version "View this project on NPM")

# grunt-node-version

> A grunt task to ensure you are using the node version required by your project's package.json

## Getting Started
This plugin requires Grunt `~0.4.1`

If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins. Once you're familiar with that process, you may install this plugin with this command:

```shell
npm install grunt-node-version --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-node-version');
```

## The "node_version" task

### Overview
In your project's Gruntfile, add a section named `node_version` to the data object passed into `grunt.initConfig()`.

```js
grunt.initConfig({
  node_version: {
    options: {
      alwaysInstall: false,
      errorLevel: 'fatal',
      globals: [],
      maxBuffer: 200*1024,
      nvm: true,
      override: ''
    }
  }
})
```

Prepend any task lists that contain [grunt-exec](https://github.com/jharding/grunt-exec) tasks with the `node_version` task, and it will extend `grunt-exec` tasks by prepending them with an `nvm use` command when necessary.

### Options

#### options.alwaysInstall
Type: `Boolean`
Default value: `false`

A boolean that determines whether to install the latest compatible version of node without a prompt (default behavior prompts user to install). This is primarily intended to be used for deployment.

#### options.errorLevel
Type: `String`
Default value: `'fatal'`

The level of error given when the wrong node version is being used. Accepted values are `'warn'` and `'fatal'`. Warn can can be overidden with `--force`, fatal cannot.

#### options.globals
Type: `Array`
Default value: `[]`

An array of node modules required to be installed globally for the project.

#### options.maxBuffer
Type: `Number`
Default value: `200*1024`

Specifies the largest amount of data allowed on stdout or stderr - if this value is exceeded then the child process is killed. If using older versions of node or NVM, you may need to increase this number for successful installation.

#### options.nvm
Type: `Boolean`
Default value: `true`

A boolean that determines whether to attempt to use/install a version of node compatible with the project using [NVM](https://github.com/creationix/nvm). If set to `false`, the plugin will just print an error if the wrong node version is being used.

#### options.override
Type: `String`
Default value: `''`

If you want to override the version specified in your project's `package.json`, specify the version you want to use instead here. This is primarily intended for testing projects on other versions of node.

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code using [Grunt](http://gruntjs.com/).

## Release History
See [History.md](https://github.com/jking90/grunt-node-version/blob/master/History.md)


History.md
## Release History
- **[1.0.0](https://github.com/jking90/grunt-node-version/tree/v1.0.0)** (Sept 5, 2013)
    - Removed `copyPackages` and `extendExec` options for lack of use case
- 0.4.1 
    - Fixed bug in older versions of NVM
- 0.4.0 
    - Removed `nvmPath` option
    - Made `~/nvm/nvm.sh` work in addition to `~/.nvm/nvm.sh`
- 0.3.1 
    - Added `override` option
- 0.3.0 
    - Added support for version ranges
- 0.2.4
    - Better logging
- 0.2.3 
    - Efficiency changes
- 0.2.2
    - Some code cleaning
- 0.2.1
    - Made more portable by switching from `~` to `process.env.HOME`
- 0.2.0
    - Added `copyPackages`, `globals` and `maxBuffer` options
    - Switched from `source` to `.`
- 0.1.1
    - Added `alwaysInstall` option
- **0.1.0**
    - First release


test/node_version_test.js
'use strict';

var grunt = require('grunt');

/*
  ======== A Handy Little Nodeunit Reference ========
  https://github.com/caolan/nodeunit

  Test methods:
    test.expect(numAssertions)
    test.done()
  Test assertions:
    test.ok(value, [message])
    test.equal(actual, expected, [message])
    test.notEqual(actual, expected, [message])
    test.deepEqual(actual, expected, [message])
    test.notDeepEqual(actual, expected, [message])
    test.strictEqual(actual, expected, [message])
    test.notStrictEqual(actual, expected, [message])
    test.throws(block, [error], [message])
    test.doesNotThrow(block, [error], [message])
    test.ifError(value)
*/

exports.node_version = {
  setUp: function(done) {
    // setup here if necessary
    done();
  },
  default_options: function(test) {
    test.expect(1);

    var actual = grunt.file.read('tmp/default_options');
    var expected = grunt.file.read('test/expected/default_options');
    test.equal(actual, expected, 'should describe what the default behavior is.');

    test.done();
  },
  custom_options: function(test) {
    test.expect(1);

    var actual = grunt.file.read('tmp/custom_options');
    var expected = grunt.file.read('test/expected/custom_options');
    test.equal(actual, expected, 'should describe what the custom option(s) behavior is.');

    test.done();
  },
};


Gruntfile.js
/*
 * grunt-node-version
 * https://github.com/jking90/grunt-node-version
 *
 * Copyright (c) 2013 Jimmy King
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>',
      ],
      options: {
        jshintrc: '.jshintrc',
      },
    },

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp'],
    },

    // Configuration to be run (and then tested).
    node_version: {
      default_options: {
        options: {
        },
        files: {
          'tmp/default_options': ['test/fixtures/testing', 'test/fixtures/123'],
        },
      },
      custom_options: {
        options: {
          separator: ': ',
          punctuation: ' !!!',
        },
        files: {
          'tmp/custom_options': ['test/fixtures/testing', 'test/fixtures/123'],
        },
      },
    },

    // Unit tests.
    nodeunit: {
      tests: ['test/*_test.js'],
    },

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-nodeunit');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'node_version', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'test']);

};


tasks/node_version.js
/*
 * grunt-node-version
 * https://github.com/jking90/grunt-node-version
 *
 * Copyright (c) 2013 Jimmy King
 * Licensed under the MIT license.
 */

'use strict';

var semver = require('semver'),
    prompt = require('prompt'),
    stripColorCodes = require('stripcolorcodes'),
    childProcess = require('child_process');

module.exports = function(grunt) {

  grunt.registerTask('node_version', 'A grunt task to ensure you are using the node version required by your project\'s package.json', function() {

    var expected = semver.validRange(grunt.file.readJSON('package.json').engines.node),
        actual = semver.valid(process.version),
        result = semver.satisfies(actual, expected),
        done = this.async(),
        home = process.env.HOME,
        locals = [],
        remotes = [],
        bestMatch = '',
        nvmUse = '',
        nvmPath = home + '/.nvm/nvm.sh',
        options = this.options({
          alwaysInstall: false,
          errorLevel: 'fatal',
          globals: [],
          maxBuffer: 200*1024,
          nvm: true,
          override: ''
        }),
        nvmInit = '. ' + nvmPath + ' && ',
        cmdOpts = {
          cwd: process.cwd(),
          env: process.env,
          maxBuffer: options.maxBuffer
        };

    // Apply override if specified
    if (options.override) {
      expected = semver.validRange(options.override);
    }

    // Extend grunt-exec
    var extendExec = function() {
      if (!result) {
        var exec = grunt.config.get('exec');
  
        for (var key in exec) {
          exec[key].cmd = nvmUse + ' && ' + exec[key].cmd;
        }
  
        grunt.config.set('exec', exec);
      }
    };

    // Validate options
    if (options.errorLevel !== 'warn' &&
        options.errorLevel !== 'fatal') {
      grunt.fail.warn('Expected node_version.options.errorLevel to be \'warn\' or \'fatal\', but found ' + options.errorLevel);
    }

    // Check for engine version in package.json
    if (!expected) {
      grunt.fail.warn('You must define a node verision in your project\'s `package.json` file.\nhttps://npmjs.org/doc/json.html#engines');
    }

    var printVersion = function(using) {
      grunt.log.write('Switched from node v' + actual + ' to ' + using);
      grunt.log.writeln('(Project requires node ' + expected + ')');
    };

    // Check for NVM
    var checkNVM = function(callback) {
      var command = '. ' + nvmPath;

      childProcess.exec(command, cmdOpts, function(err, stdout, stderr) {
        if (stderr.indexOf('No such file or directory') !== -1) {
          if (nvmPath === home + '/.nvm/nvm.sh') {
            nvmPath = home + '/nvm/nvm.sh';
            nvmInit = '. ' + nvmPath + ' && ';
            checkNVM(callback);
          } else {
            grunt[options.errorLevel]('Expected node ' + expected + ', but found v' + actual + '\nNVM does not appear to be installed.\nPlease install (https://github.com/creationix/nvm#installation), or update the NVM path.');
          }
        } else {
          callback();
        }
      });
    };

    // Check for globally required packages
    var checkPackages = function (packages) {
      var thisPackage;

      if (packages.length) {
        thisPackage = packages.pop();

        var command = nvmUse + ' && npm ls -g ' + thisPackage;

        childProcess.exec(command, cmdOpts,function(err, stdout, stderr) {
          if (err) { throw err ;}

          if (stdout.indexOf('â”€ (empty)') !== -1) {
            npmInstall(thisPackage, function() {
              checkPackages(packages);        
            });
          } else {
            checkPackages(packages);
          }
        });

      } else {
        done();
      }
    };

    // Install missing packages
    var npmInstall = function(thisPackage, callback) {
      var command = nvmUse + ' && npm install -g ' + thisPackage;

      childProcess.exec(command, cmdOpts,function(err, stdout, stderr) {
        if (err) { throw err ;}
        grunt.verbose.writeln(stdout);
        grunt.log.oklns('Installed ' + thisPackage);
        callback();
      });
    };

    // Prompt to install
    var askInstall = function() {
      prompt.start();

      var prop = {
        name: 'yesno',
        message: 'You do not have any node versions installed that satisfy this project\'s requirements ('.white + expected.yellow + '). Would you like to install the latest compatible version? (y/n)'.white,
        validator: /y[es]*|n[o]?/,
        required: true,
        warning: 'Must respond yes or no'
      };

      prompt.get(prop, function (err, result) {
        result = result.yesno.toLowerCase();
        if (result === 'yes' ||
            result === 'y') {
          nvmInstall();
        } else {
          grunt[options.errorLevel]('Expected node v' + expected + ', but found ' + actual);
        }
      });
    
    };

    // Install latest compatible node version
    var nvmInstall = function() {
      nvmLs('remote', function() {

        bestMatch = semver.maxSatisfying(remotes, expected);
        nvmUse = nvmInit + 'nvm use ' + bestMatch;

        var command = nvmInit + 'nvm install ' + bestMatch;

        childProcess.exec(command, cmdOpts,function(err, stdout, stderr) {
          if (err) { throw err ;}
          var nodeVersion = stdout.split(' ')[3];
          grunt.log.ok('Installed node v' + bestMatch);
          printVersion(nodeVersion);
          extendExec();
          checkPackages(options.globals);
        });
      });
    };

    // Check for available node versions
    var nvmLs = function(loc, callback) {
      var command = nvmInit + 'nvm ls';

      if (loc === 'remote') {
        command += '-remote';
      }

      childProcess.exec(command, cmdOpts, function(err, stdout, stderr) {
        var data = stripColorCodes(stdout.toString()).replace(/\s+/g, '|'),
            available = data.split('|');

        for (var i = 0; i < available.length; i++) {
          // Trim whitespace
          available[i] = available[i].replace(/\s/g, '');
          // Validate
          var ver = semver.valid(available[i]);
          if (ver) {
            if (loc === 'remote') {
              remotes.push(ver);
            } else if (loc === 'local') {
              locals.push(ver);
            }
          }
        }
        
        callback();
      });
    };

    // Check for compatible node version
    var checkVersion = function() {
      // Make sure a node version is intalled that satisfies
      // the projects required engine. If not, prompt to install.
      nvmLs('local', function() {
        var matches = semver.maxSatisfying(locals, expected);

        if (matches) {
          bestMatch = matches;
          nvmUse = nvmInit + 'nvm use ' + bestMatch;

          childProcess.exec(nvmUse, cmdOpts,function(err, stdout, stderr) {
            printVersion(stdout.split(' ')[3]);
            extendExec();
            checkPackages(options.globals);
          });
        } else {
          if (options.alwaysInstall) {
            nvmInstall();
          } else {
            askInstall();
          }
        }
      });
    };

    if (result === true) {
      grunt.log.writeln('Using node ' + actual);
      grunt.log.writeln('(Project requires node ' + expected + ')');
      checkPackages(options.globals);
    } else {
      if (!options.nvm) {
        grunt[options.errorLevel]('Expected node ' + expected + ', but found v' + actual);
      } else {
        checkNVM(checkVersion);
      }
    }

  });

};
